docs/react-jss-hoc-migration.md

```diff
@@ -0,0 +1,134 @@
+# Migrating the `withStyles` HOC
+
+Although we recommend using the new hooks, it's possible that you have class components that cannot be migrated easily at this time. In that case, you can create your own higher order component (HOC) from the provided hooks. This way, you'll have a HOC that stays up-to-date with the latest features, and you'll still have the option of fully migrating to hooks at your own convenience.
+
+A simple solution may look something like this:
+
+```jsx
+import React from 'react'
+import {createUseStyles, useTheme} from 'react-jss'
+
+/**
+ * Creates a Higher Order Component that injects the CSS specified in `styles`.
+ * @param styles
+ */
+function withStyles(styles) {
+  return function(WrappedComponent) {
+    const useStyles = createUseStyles(styles)
+
+    const StyledComponent = props => {
+      const {classes, ...passThroughProps} = props
+      const theme = useTheme()
+      const reactJssClasses = useStyles({...passThroughProps, theme})
```

Comment:
there is also merging of classes happening right now in withStyles, I am wondering if we should just rewrite the HOC to use hooks like this and make it a separate entry point so that it doesn't increase bundle size for everyone even if one doesn't have tree shaking or alternatively just link them to the latest version current implementation that they can just copy it into their code base as is and not bother creating their own potentially incomplete versions.

Since we are not going to have to fix it any more, people will only need it to keep the current code working and use hooks directly for the new code.

Reply:
Potentially we would need to make it copy/pastable because currently withStyles reuses quite a bit code with hooks

Reply:
Just thought about it a bit more, I think we should start splitting react-jss into multiple packages:

- hooks 
- hoc
- theming
- css() (unofficial yet) 
- styled() (unofficial yet)
- JssProvider (only needed for customization and different setups for subtrees or SSR won't be needed by many people or only in one place)


By doing so we can keep them using the hocs interface from a separate package as it is now many more years, since it's not increasing bundle size

react-jss also has a generally big bundle size because of how bundlephobia doesn't know what is actually used, so it can not treeshake

by splitting into packages, users will know what exactly they are using and will see specific packages in bundlephobia

This is slightly off-top here, but if we go this route, we don't need to explain how to create own withStyles HOC, so it matters

I am also thinking that creating a new preset package with only plugins which most people expect to be inside, something comparable to emotion/SC from capabilities perspective, will largely improve bundle size and hence perception of the library

In addition to that a few things can be removed from JSS that haven't been properly documented or used anyways and a few files can be entirely moved to a separate package so that jss core can also have a smaller bundle size

Reply:
At the moment, I don't have time to respond to these comments, but I will respond later.

Just putting that out there since I responded to the other comment. Didn't want it to seem like I was ignoring the ones you gave here. 😅